### PR TITLE
Week5 (#24)

### PR DESCRIPTION
**Pull request by lyabomyr**
* week5 seldon, fastapi, streamlit

### CODE REVIEW

File: .github/workflows/week5.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      - name: set up docker build
+        uses: docker/setup-buildx-action@v2
+
+      - name: Build and push
+        uses: docker/build-push-action@v3
+        with:
+          context: .
+          file:  week5/Dockerfile
+          push: true
+          tags: ${{secrets.LOGIN_DOCKER_HUB}}/${{env.IMAGE_SIMLESERVER}}:latest
```
Comment (truskovskiyk) [inline .github/workflows/week5.yml]: where is test step?

File: week5/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+ENV WANDB_API_KEY="0a14fd83112d8258ad9cb780edb9e9932241f5fc"
+
+ENV PYTHONPATH="$PWD/ml_project"
+COPY week5/serving /ml_project/serving
+COPY week5/unit_tests /ml_project/unit_tests
+COPY Churn_Modelling.csv /ml_project/Churn_Modelling.csv
+COPY week5/test_churn_modelling.csv /ml_project/test_churn_modelling.csv
+
+RUN echo $PYTHONPATH
+RUN pytest unit_tests/test_fastapi.py
```
Comment (truskovskiyk) [inline week5/Dockerfile]: let's not call this here, and call in CI as separate step, independent from building process

File: week5/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,52 @@
+FROM python:latest as base
```
Comment (truskovskiyk) [inline week5/Dockerfile]: `python:latest` -> let's use defined version of python instead latest

File: week5/README.txt [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+curl -fsSL https://get.deta.dev/cli.sh | sh
+```
+
+ open new shell and put command in derictiry with main.pt file
+ ```
+ deta --help
+ dete login
+ deta new
+ ```
+ open endpoint URL from json in terminal
```
Comment (truskovskiyk) [inline week5/README.txt]: interesting

File: week5/README.txt [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    --namespace seldon-system
+```
+Port Forward:
+```
+kubectl port-forward  --address 0.0.0.0 -n ambassador svc/ambassador 7777:80
+```
+
+run custom yaml file:
+go to swager
+http://localhost:7777/seldon/default/churnpredict/api/v1.0/doc/#/
```
Comment (truskovskiyk) [inline week5/README.txt:42]: should we deploy model before go to `http://localhost:7777/seldon/default/churnpredict/api/v1.0/doc/#/`?

File: week5/main.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+
+@app.get("/")
+def read_root():
+    return {"Hello": "please add \'/docs\' to ur url and u will be transferred to swagger page with list of request "
+                     "example url http://127.0.0.1:8000/docs"}
+
+
+@app.post("/train_model_and_push_to_wandb/")
+async def train_model_and_push_to_wandb(file: UploadFile = File(...)):
```
Comment (truskovskiyk) [inline week5/main.py:44]: should we specify returning scheme here?

Comment (truskovskiyk) [inline week5/main.py:44]: ooo, wow you train the model on request, nice!

Comment (truskovskiyk) [inline week5/main.py:44]: why async? this is CPU bound function, let's use just normal function

File: week5/main.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        df = request_csv_to_df(file)
+        train = grid_train_model(prepare_data_for_train(df)[0]['x'], prepare_data_for_train(df)[0]['y'],
+                                 model_file_name)
+        model_loader('upload', model_file_name)
+        clean_all(model_file_name)
+        return {"Complete train and sent to wandb with model_name": train}
+
+
+@app.post("/batch_predict_csv_file/")
+async def predict_exel_file(file: UploadFile = File(...)):
```
Comment (truskovskiyk) [inline week5/main.py:57]: why async?

Comment (truskovskiyk) [inline week5/main.py:57]: let's add response schema here

File: week5/serving/.streamlit/config.toml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+[theme]
+primaryColor="#F63366"
+backgroundColor="#FFFFFF"
+secondaryBackgroundColor="#F0F2F6"
+textColor="#262730"
+font="sans serif"
```
Comment (truskovskiyk) [inline week5/serving/.streamlit/config.toml:6]: nice! customization

File: week5/serving/fast_api.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
+from fastapi import FastAPI, File, UploadFile
```
Comment (truskovskiyk) [inline week5/serving/fast_api.py:3]: hm, I already saw file `[main.py](https://github.com/lyabomyr/Projector_ML_in_production_course/pull/25/files#diff-9fa53c587e8ce459bf9a8eb0926e7da1d1722984804857427f60fa9e54382d7b)`

is it duplicated file?

File: week5/serving/ml_decision_tree_sample/cleaner.py [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+        print(f"The file {model_name} has been deleted successfully")
+    else:
+        print(f"The file {model_name} does not exist!")
+
+
+def remove_wand_local_directory():
+    try:
+        os.system("rm -rf wandb/")
+    except:
+        pass
```
Comment (truskovskiyk) [inline week5/serving/ml_decision_tree_sample/cleaner.py]: don't use `except:`

File: week5/serving/ml_decision_tree_sample/cleaner.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    if os.path.exists(model_name):
+        os.remove(model_name)
+        print(f"The file {model_name} has been deleted successfully")
+    else:
+        print(f"The file {model_name} does not exist!")
+
+
+def remove_wand_local_directory():
+    try:
+        os.system("rm -rf wandb/")
```
Comment (truskovskiyk) [inline week5/serving/ml_decision_tree_sample/cleaner.py]: don't use `os.system("rm -rf wandb/")`

please use: `shutil.rmtree`

File: week5/serving/ml_decision_tree_sample/load_model.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+            artifact.add_file(model_name)
+            run.log_artifact(artifact)
+        else:
+            artifact = run.use_artifact(f'{model_name}:latest', type="model")
+            artifact = artifact.get_path(model_name)
+            artifact.download(root='.')
+
+
+
+# model_loader('download','model_random_forest')
```
Comment (truskovskiyk) [inline week5/serving/ml_decision_tree_sample/load_model.py]: remove commented code

File: week5/serving/seldon_api.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+from serving.config import model_file_name
+import logging
+
+logger = logging.getLogger()
+
+
+class SeldonAPI:
+    def __init__(self):
+        model_loader('download', model_file_name)
+        print(model_file_name)
```
Comment (truskovskiyk) [inline week5/serving/seldon_api.py]: interesting

**Review by truskovskiyk — COMMENTED** [review]
@lyabomyr amazing start! please see feedback!

File: week5/serving/seldon_api.py (hunk 13 above)
Comment (lyabomyr) [reply week5/serving/seldon_api.py]: Removed print

File: week5/serving/ml_decision_tree_sample/load_model.py (hunk 12 above)
Comment (lyabomyr) [reply week5/serving/ml_decision_tree_sample/load_model.py]: removed

File: week5/serving/ml_decision_tree_sample/cleaner.py (hunk 11 above)
Comment (lyabomyr) [reply week5/serving/ml_decision_tree_sample/cleaner.py]: did it

File: week5/serving/ml_decision_tree_sample/cleaner.py (hunk 10 above)
Comment (lyabomyr) [reply week5/serving/ml_decision_tree_sample/cleaner.py]: Changed
But why don't use?

File: week5/serving/fast_api.py (hunk 9 above)
Comment (lyabomyr) [reply week5/serving/fast_api.py:3]: file main.py I need for deployment to . deta  (https://fastapi.tiangolo.com/deployment/deta/)

File: week5/main.py (hunk 6 above)
Comment (lyabomyr) [reply week5/main.py:44]: train request nothing to return

File: week5/Dockerfile (hunk 3 above)
Comment (lyabomyr) [reply week5/Dockerfile]: fixed

File: week5/Dockerfile (hunk 2 above)
Comment (lyabomyr) [reply week5/Dockerfile]: ok, fixed

File: .github/workflows/week5.yml (hunk 1 above)
Comment (lyabomyr) [reply .github/workflows/week5.yml]: added

File: week5/main.py (hunk 7 above)
Comment (lyabomyr) [reply week5/main.py:57]: removed

File: week5/main.py (hunk 6 above)
Comment (lyabomyr) [reply week5/main.py:44]: removed

File: week5/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+LABEL maintainer="lyabomyr@gmail.com"
+RUN pip install --upgrade pip
+
+WORKDIR /ml_project
+COPY week5/requirements.txt /ml_project/requirements.txt
+RUN pip install --no-cache-dir --upgrade -r  /ml_project/requirements.txt
+
+
+ENV WANDB_PROJECT="liunomyr"
+ENV WANDB_API_KEY="0a14fd83112d8258ad9cb780edb9e9932241f5fc"
```
Comment (truskovskiyk) [inline week5/Dockerfile]: remove cred from github

File: week1/k8s-resources/job-ml.yaml [hunk 15]
```diff
@@ -1,4 +1,4 @@
-apiVersion: batch/v1
+xapiVersion: batch/v1
```
Comment (truskovskiyk) [inline week1/k8s-resources/job-ml.yaml:1]: hm, not sure here is such thing as `xapiVersion`

File: week5/unit_tests/test_benchmark-forwardpass.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+    result = input_df
+    return result
+
+
+def test_load_model(benchmark):
+    benchmark(download_model)
+
+
+def test_benchmark_port_forwarding_batch_predict(benchmark):
+    benchmark(my_predict)
```
Comment (truskovskiyk) [inline week5/unit_tests/test_benchmark-forwardpass.py]: could you add results here

File: week5/unit_tests/test_benchmark_optimization_inference.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+            futures.append(future)
+        wait(futures)
+        y_pred = []
+        for future in futures:
+            y_batch = future.result()
+            y_pred.append(y_batch)
+    return np.concatenate(y_pred)
+
+def test_optimized_predict(benchmark):
+    benchmark(run_inference_process_pool)
```
Comment (truskovskiyk) [inline week5/unit_tests/test_benchmark_optimization_inference.py:53]: this is not optimization, please look at the work for reference what model optimization is: 

https://awesomeopensource.com/projects/model-compression
https://arxiv.org/abs/1710.09282
https://docs.google.com/document/d/1hNuW6bqWYZjlwpit_8W1cu7kllb-jTfy3Liof1GJWug/edit

File: week1/k8s-resources/job-ml.yaml (hunk 15 above)
Comment (lyabomyr) [reply week1/k8s-resources/job-ml.yaml:1]: my bad

Comment (lyabomyr) [reply week1/k8s-resources/job-ml.yaml:1]: Fixed

File: week5/unit_tests/test_benchmark-forwardpass.py (hunk 16 above)
Comment (lyabomyr) [reply week5/unit_tests/test_benchmark-forwardpass.py]: Result was added in readme.md
I will add to this code like comment to

File: week5/unit_tests/test_benchmark_optimization_inference.py (hunk 17 above)
Comment (lyabomyr) [reply week5/unit_tests/test_benchmark_optimization_inference.py:53]: But I believe this good optimization too)))

File: week5/unit_tests/test_benchmark-forwardpass.py (hunk 16 above)
Comment (lyabomyr) [reply week5/unit_tests/test_benchmark-forwardpass.py]: Added result like comment

Comment (lyabomyr) [reply week5/unit_tests/test_benchmark-forwardpass.py]: Added result like comments

**Review by truskovskiyk — APPROVED** [review]

**Pull request merged**